Dockerfile

```diff
@@ -0,0 +1,24 @@
+FROM golang:1.18 as builder
+
+WORKDIR /workspace
+# Copy the Go Modules manifests
+COPY go.mod go.mod
+COPY go.sum go.sum
+# cache deps before building and copying source so that we don't need to re-download as much
+# and so that source changes don't invalidate our downloaded layer
+RUN go mod download
+
+# Copy the go source
+COPY . /workspace
+
+# Build
+RUN make build
```

Comment:
Why build here again but not use artifact from build and test?

Reply:
keeping the `Dockerfile` mutually exclusive from the CI flow so a
 * dev can run `make docker-build` or `docker build .` to build images and test them quickly
 * above flow eliminates volume mounts to share artifacts etc

Reply:
I feel this is a bad practice: we are releasing a build which is not tested in the same environment.

buildx doesn't need volume mounts anyway, it is just `ADD` file?

Reply:
I disagree here, the workflow for pushing docker images would also run some form of e2e (which doesn't exist today) using the same image before pushing the image,

Reply:
The other disadvantage of this is we are going to use QEMU for future arm64 (and other) build which is way slower and hard to workaround. It is better to split the build and docker packaging.

Reply:
prefer if we first tried the simpler way of building `linux/arm64` using QEMU, quantified the slowness in CI, and then deleted the QEMU step in favor of a self hosted ARM runner

Reply:
You don't need even a self hosted ARM runner, Go cross compile works great without hassle, just need to copy the binary from local machine (or downloaded artifacts), instead of force docker build run Go build.

Reply:
raised https://github.com/envoyproxy/gateway/issues/93 so this conversation can continue there and can unblock this PR